### PR TITLE
Adddlatrd

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

File: lapack.go [hunk 1]
```diff
@@ -100,6 +100,15 @@ const (
 	ApplyQ DecompUpdate = 'Q'
 )
 
+// EigComp specifies the type of eigenvalue decomposition.
+type EigComp byte
```
Comment (vladimir-ch) [inline lapack.go]: Where is `EigComp` used? I cannot find it.

Comment (btracey) [inline lapack.go]: In a future commit apparently. Removed.

**Comment by btracey** [comment]
PTAL

**Comment by kortschak** [comment]
badEigComp should probably go away too then.

File: native/dlatrd.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+			if i > 0 {
+				// Generate elementary reflector H(i) to annihilate A(0:i-2,i).
+				a[(i-1)*lda+i], tau[i-1] = impl.Dlarfg(i, a[(i-1)*lda+i], a[i:], lda)
+				e[i-1] = a[(i-1)*lda+i]
+				a[(i-1)*lda+i] = 1
+
+				// Compute W(0:i-1, i).
+				bi.Dsymv(blas.Upper, i, 1, a, lda, a[i:], lda, 0, w[iw:], ldw)
+				if i < n-1 {
+					bi.Dgemv(blas.Trans, i, n-i-1, 1, w[iw+1:], ldw, a[i:], lda, 0, w[(i+1)*ldw+iw:], ldw)
```
Comment (kortschak) [inline native/dlatrd.go]: Split these over lines - purely because this block is hard to read otherwise.

File: native/dlatrd.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+		}
+	} else {
+		// Reduce first nb columns of lower triangle.
+		for i := 0; i < nb; i++ {
+			// Update A(i:n, i)
+			bi.Dgemv(blas.NoTrans, n-i, i, -1, a[i*lda:], lda, w[i*ldw:], 1, 1, a[i*lda+i:], lda)
+			bi.Dgemv(blas.NoTrans, n-i, i, -1, w[i*ldw:], ldw, a[i*lda:], 1, 1, a[i*lda+i:], lda)
+			if i < n-1 {
+				// Generate elementary reflector H(i) to annihilate A(i+2:n,i).
+				a[(i+1)*lda+i], tau[i] = impl.Dlarfg(n-i-1, a[(i+1)*lda+i], a[min(i+2, n-1)*lda+i:], lda)
```
Comment (kortschak) [inline native/dlatrd.go]: Seems to me that you can directly assign to e[i], no?

File: native/dlatrd.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+				e[i] = a[(i+1)*lda+i]
+				a[(i+1)*lda+i] = 1
+
+				// Compute W(i+1:n,i).
+				bi.Dsymv(blas.Lower, n-i-1, 1, a[(i+1)*lda+i+1:], lda, a[(i+1)*lda+i:], lda, 0, w[(i+1)*ldw+i:], ldw)
```
Comment (kortschak) [inline native/dlatrd.go]: Split these over lines.

File: native/dlatrd.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+//
+// w is an n×nb matrix. On exit it contains the data to update the unreduced part
+// of A.
+//
+// The matrix Q is represented as a product of elementary reflectors. Each reflector
+// H has the form
+//  I - tau * v * v^T
+// If uplo == blas.Upper,
+//  Q = H[n] * H[n-1] * ... * H[n-nb+1]
+// where v[:i-1] is stored in A[:i-1,i], v[i-1] = 1, and v[i:n] = 0
```
Comment (vladimir-ch) [inline native/dlatrd.go]: Final dot.

File: native/dlatrd.go (hunk 2 above)
Comment (btracey) [inline native/dlatrd.go]: Done.

File: native/dlatrd.go (hunk 3 above)
Comment (btracey) [inline native/dlatrd.go]: Good call.

File: native/dlatrd.go (hunk 4 above)
Comment (btracey) [inline native/dlatrd.go]: Done.

File: native/dlatrd.go (hunk 5 above)
Comment (btracey) [inline native/dlatrd.go]: Done.

File: native/dlatrd.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+// reduced. If uplo == blas.Lower, the lower triangle is supplied and the first
+// nb rows are reduced.
+//
+// a contains the symmetric matrix on entry with active triangular half specified
+// by uplo. On exit, the nb columns have been reduced to tridiagonal form. The
+// diagonal contains the diagonal of the reduced matrix, the off-diagonal is
+// set to 1, and the remaining elements contain the data to construct Q.
+//
+// If uplo == blas.Upper, with n = 5 and nb = 2 on exit a is
+//  [a  a  a v4 v5]
```
Comment (vladimir-ch) [inline native/dlatrd.go]: In utf8, there is a nice set of symbols for denoting matrices like these:
```
⎡
⎢
⎣
```
and
```
⎤
⎥
⎦
```
Just saying.

Comment (btracey) [inline native/dlatrd.go]: Woah! Fixed in this function.

File: native/dlatrd.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+		for i := n - 1; i >= n-nb; i-- {
+			iw := i - n + nb
+			if i < n-1 {
+				// Update A(0:i, i).
+				bi.Dgemv(blas.NoTrans, i+1, n-i-1, -1, a[i+1:], lda, w[i*ldw+iw+1:], 1, 1, a[i:], lda)
+				bi.Dgemv(blas.NoTrans, i+1, n-i-1, -1, w[iw+1:], ldw, a[i*lda+i+1:], 1, 1, a[i:], lda)
+			}
+			if i > 0 {
+				// Generate elementary reflector H(i) to annihilate A(0:i-2,i).
+				a[(i-1)*lda+i], tau[i-1] = impl.Dlarfg(i, a[(i-1)*lda+i], a[i:], lda)
```
Comment (vladimir-ch) [inline native/dlatrd.go]: Directly assign to `e` also here?

Comment (btracey) [inline native/dlatrd.go]: Yea, got that one too

File: testlapack/dlatrd.go [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+				}
+			}
+		}
+	}
+	return true
+}
+
+// checkOrthonormal checks that a general matrix is orthonormal.
+// TODO(btracey): Replace other tests with a call to this function.
+func isOrthonormal(t *testing.T, q blas64.General) bool {
```
Comment (vladimir-ch) [inline testlapack/dlatrd.go]: It does not need `t`.

File: testlapack/dlatrd.go [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+						return false
+					}
+				}
+			}
+		}
+	}
+	return true
+}
+
+// checkOrthonormal checks that a general matrix is orthonormal.
```
Comment (vladimir-ch) [inline testlapack/dlatrd.go]: s/checkOrthonormal/isOrtonormal/

Comment (btracey) [inline testlapack/dlatrd.go]: Done.

File: testlapack/dlatrd.go (hunk 8 above)
Comment (btracey) [inline testlapack/dlatrd.go]: Done.

File: native/dlatrd.go [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+//
+// If uplo == blas.Upper, with n = 5 and nb = 2 on exit a is
+//  ⎡a  a  a v4 v5]
+//  ⎢   a  a v4 v5]
+//  ⎢      a  1 v5]
+//  ⎢         d  1]
+//  ⎣            d]
+//
+// If uplo == blas.Lower, with n = 5 and nb = 2, on exit a is
+//  ⎡d            ⎤
```
Comment (vladimir-ch) [inline native/dlatrd.go]: Also the right bracket.

Comment (vladimir-ch) [inline native/dlatrd.go]: I mean the matrix above.

Comment (btracey) [inline native/dlatrd.go]: Yea, got it. Sorry.

**Comment by btracey** [comment]
PTAL

File: native/dlatrd.go (hunk 6 above)
Comment (kortschak) [inline native/dlatrd.go]: Unfortunately they don't always render properly (looking at you, ios).

Comment (btracey) [inline native/dlatrd.go]: Still better than lots of brackets though?

File: native/dlatrd.go [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+// reduced. If uplo == blas.Lower, the lower triangle is supplied and the first
+// nb rows are reduced.
+//
+// a contains the symmetric matrix on entry with active triangular half specified
+// by uplo. On exit, the nb columns have been reduced to tridiagonal form. The
+// diagonal contains the diagonal of the reduced matrix, the off-diagonal is
+// set to 1, and the remaining elements contain the data to construct Q.
+//
+// If uplo == blas.Upper, with n = 5 and nb = 2 on exit a is
+//  ⎡a  a  a v4 v5⎤
```
Comment (kortschak) [inline native/dlatrd.go]: I know that I used these in mat64.Format (note that there is still an issue to make that more general and allow ASCII-only formatting), but please don't use them in godoc; they don't render correctly on a non-zero set of machines.

Comment (btracey) [inline native/dlatrd.go]: Okay, undone.

File: native/dlatrd.go (hunk 6 above)
Comment (kortschak) [inline native/dlatrd.go]: On iOS (here) they render as the unknown glyph which is just a box.

Comment (btracey) [inline native/dlatrd.go]: Reverted

**Comment by btracey** [comment]
Is this ready?

**Comment by vladimir-ch** [comment]
LGTM but wait for @kortschak

**Comment by kortschak** [comment]
LGTM

**Comment by btracey** [comment]
Woo! One more down. Getting closer...

**Pull request merged**